include/clock_control.h

```diff
@@ -72,6 +131,66 @@ static inline int clock_control_off(struct device *dev,
 	return api->off(dev, sys);
 }
 
+/**
+ * @brief Request clock to start with notification when clock has been started.
+ *
+ * User can request delayed start by providing exact information when clock
+ * should be ready. Driver ensures that clock is ready before requested time.
+ * It is the driver responsibility to take into account clock startup time.
+ * When clock is already running user callback will be called from the context
+ * of the function call else it is called from other context (e.g. clock
+ * interrupt).
+ *
+ * @param dev 	   Device.
+ * @param sys	   A pointer to an opaque data representing the sub-system.
+ * @param data	   Data structure containing a callback that is called when
+ *		   action is performed. Structure content must be valid until
+ *		   clock is started and user callback is called. Can be NULL.
+ * @param deadline When (given in @ref k_cycle_get_32()) clock must be ready.
+ *		   Requested time must be no later than 2^31 cycles from the
+ *		   moment it is scheduled. Other values are considered as
+ *		   immediate request. Use @ref k_cycle_get_32() for immediate
+ *		   request.
+ *
+ * @retval 0 if clock is started or already running.
+ * @retval -EBUSY if same request already scheduled and not yet completed.
+ * @retval -ENOTSUP if not supported.
+ */
+static inline int clock_control_async_on(struct device *dev,
+					 clock_control_subsys_t sys,
+					 struct clock_control_async_data *data,
+					 u32_t deadline)
+{
+	const struct clock_control_driver_api *api =
+		(const struct clock_control_driver_api *)dev->driver_api;
+
+	if (!api->async_on) {
+		return -ENOTSUP;
+	}
+
+	return api->async_on(dev, sys, data, deadline);
+}
+
+/**
+ * @brief Get clock status.
+ *
+ * @param dev Device.
+ *
+ * @return Status.
+ */
+static inline enum clock_control_status clock_control_get_status(
+							struct device *dev)
+{
+	const struct clock_control_driver_api *api =
+		(const struct clock_control_driver_api *)dev->driver_api;
+
+	if (!api->get_status) {
+		return CLOCK_CONTROL_STATUS_UNKNOWN;
```

Comment:
enum used here?

We want to mix positive enum and negative errno?

Reply:
but this function does not return error codes only enum.

Reply:
my fault, i assumed the return value was of type int in the signature, clearly it is enum. Resolved.